### PR TITLE
feat: EMA for unequally sized time series

### PR DESCRIPTION
**Pull request by 3Hren**

### CODE REVIEW

File: include/metrics/usts/ewma.hpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    /// By initializing moving average with its first value added we introduce an error, which
+    /// declines respectively for a discrete series. After the build-up time, when we start using
+    /// the EWMA, we expect it to be (almost) free of initialization errors.
+    auto warmed_up() const -> bool;
+
+    /// Adds a point updating the moving average.
+    ///
+    /// \param time Time point. Must be monotonically increased comparing with previous values.
+    /// \param value Observed value.
+    auto add(time_point time, double value) -> void;
```
Comment (antmat) [inline include/metrics/usts/ewma.hpp:66]: I think you can also add 
`auto add(double value) -> void;`
 overload to add point for current time.

Besides adding with timepoint can "spoil" metric if somebody adds an event deeply in future.

File: src/usts/ewma.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+auto ewma<Clock>::add(time_point time, double value) -> void {
+    if (time < prev) {
+        throw std::invalid_argument("time argument must monotonically increase");
+    }
+
+    if (initialized.test_and_set()) {
+        const auto ẟ = std::chrono::duration_cast<std::chrono::nanoseconds>(time - prev).count();
+        const auto α = ẟ / τ;
+        const auto µ = std::exp(-α);
+        average = µ * average + (1 - µ) * value;
```
Comment (antmat) [inline src/usts/ewma.cpp]: Looks valid, but I don't fully understand where does it go from.

**Comment by 3Hren** [comment]
Something like this:
Here (τ=20):
 - Blue line - real value.
 - Red line - exact EMA calculated using prime formula.
 - Green line - EMA, calculated using recursive formula. Initial divergence is explained with initialization error.

image

File: include/metrics/usts/ewma.hpp (hunk 1 above)
Comment (3Hren) [reply include/metrics/usts/ewma.hpp:66]: Agree.

**Comment by 3Hren** [comment]
@borislitv @ijon @toshik you'd also may be interested.

**Pull request merged**